WIP

dongrichain opened the pull request:
yui-ibc-solidityのバージョンを最新に反映してみました。
変更した理由: 
1. Minitokenが最新バージョンのyui-ibc-solidityに適用してなかった
2. 古いバージョンだとnode 18でnpm installしようとすると依存関係のlibが古くてinstallできない
![commonjs](https://user-images.githubusercontent.com/126585618/225823798-bef6528f-a2db-492f-b2ac-c36ebe53c492.png)

# 変更内容
1. package.jsonで"@hyperledger-labs/yui-ibc-solidityのgit参照先を最新に変更
2. protoからコード生成時のimport path修正
3. MintoToken.solのimport path, 処理一部修正
4. migrationで使わなくなったIBCHost削除とMiniTokenデプロイ周り修正

# 結果
![スクリーンショット 2023-03-17 14 57 28](https://user-images.githubusercontent.com/126585618/225824449-1b13b30f-cac7-4b4a-82ea-7c87758f1528.png)
このようにtransactionが失敗して、接続できない

siburu commented:
create client, update clientには成功しており、その後のconnection handshakeの最初のステップであるconnOpenInitで失敗しているようですね

dongrichain commented:
transfer時ibc1で受信できない問題

ibc0側でaliceからbobに50トークン転送コマンド実行、txhashでイベントも正常
```
const port = "transfer";
const channel = "channel-0";

const bob = accounts[2];
await MiniToken.deployed()
    .then(instance => instance.sendTransfer(50, bob, port, channel, 0, {from: alice}));
```

## log
```
truffle(ibc0)> await MiniToken.deployed().then(instance => instance.sendTransfer(50, bob, port, channel, 0, {from: alice}));
{
  tx: '0x7565f30e336d0081f45c87339b1ed3cd3c7e3930756732ea4c42d6219f18a7ee',
  receipt: {
    blockHash: '0xd0f1f9e309cc1642fcc60f393ba97c73d71868686a7679d067fee815be34f7e8',
    blockNumber: 200,
    contractAddress: null,
    cumulativeGasUsed: 154453,
    from: '0xcbed645b1c1a6254f1149df51d3591c6b3803007',
    gasUsed: 154453,
    logs: [ [Object], [Object] ],
    logsBloom: '0x00000000002000000000000000000000000000000000001800000000000002000000000000008000000000000000000000004000000000000000000000004000000000000004200000000000000000000000200000000000000000000000000000000000000000000000000000008000080000010000000000000000000000000000000000000000000000000000000000000000000000000000000000600002000000000000000000000000000000000000800000000000000000000000000004000000000000000000000000000000000000000000000000800000000000000000000000001000000000000000000000200000000000000000000000000000',
    status: true,
    to: '0xae1c9125bbcf63bf51294c4d15cbd472782e330d',
    transactionHash: '0x7565f30e336d0081f45c87339b1ed3cd3c7e3930756732ea4c42d6219f18a7ee',
    transactionIndex: 0,
    rawLogs: [ [Object], [Object], [Object] ]
  },
  logs: [
    {
      address: '0xaE1C9125BbcF63bf51294C4D15CBD472782E330D',
      blockNumber: 200,
      transactionHash: '0x7565f30e336d0081f45c87339b1ed3cd3c7e3930756732ea4c42d6219f18a7ee',
      transactionIndex: 0,
      blockHash: '0xd0f1f9e309cc1642fcc60f393ba97c73d71868686a7679d067fee815be34f7e8',
      logIndex: 0,
      removed: false,
      id: 'log_1e24d210',
      event: 'Burn',
      args: [Result]
    },
    {
      address: '0xaE1C9125BbcF63bf51294C4D15CBD472782E330D',
      blockNumber: 200,
      transactionHash: '0x7565f30e336d0081f45c87339b1ed3cd3c7e3930756732ea4c42d6219f18a7ee',
      transactionIndex: 0,
      blockHash: '0xd0f1f9e309cc1642fcc60f393ba97c73d71868686a7679d067fee815be34f7e8',
      logIndex: 2,
      removed: false,
      id: 'log_e7ed5c09',
      event: 'SendTransfer',
      args: [Result]
    }
  ]
}

truffle(ibc0)> MiniToken.deployed().then(instance => instance.getPastEvents("SendTransfer", { fromBlock: 0 })).then(event => console.log(event));
[
  {
    logIndex: 2,
    removed: false,
    blockNumber: 200,
    blockHash: '0xd0f1f9e309cc1642fcc60f393ba97c73d71868686a7679d067fee815be34f7e8',
    transactionHash: '0x7565f30e336d0081f45c87339b1ed3cd3c7e3930756732ea4c42d6219f18a7ee',
    transactionIndex: 0,
    address: '0xaE1C9125BbcF63bf51294C4D15CBD472782E330D',
    id: 'log_e7ed5c09',
    returnValues: Result {
      '0': '0xcBED645B1C1a6254f1149Df51d3591c6B3803007',
      '1': '0x00731540cd6060991D6B9C57CE295998d9bC2faB',
      '2': 'transfer',
      '3': 'channel-0',
      '4': '0',
      '5': '50',
      from: '0xcBED645B1C1a6254f1149Df51d3591c6B3803007',
      to: '0x00731540cd6060991D6B9C57CE295998d9bC2faB',
      sourcePort: 'transfer',
      sourceChannel: 'channel-0',
      timeoutHeight: '0',
      amount: '50'
    },
    event: 'SendTransfer',
    signature: '0x09a4832cd64f583d0bed9b5b606965217862ef4de4c4b54e8b89cff6225dca02',
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003200000000000000000000000000000000000000000000000000000000000000087472616e7366657200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096368616e6e656c2d300000000000000000000000000000000000000000000000',
      topics: [Array]
    },
    args: Result {
      '0': '0xcBED645B1C1a6254f1149Df51d3591c6B3803007',
      '1': '0x00731540cd6060991D6B9C57CE295998d9bC2faB',
      '2': 'transfer',
      '3': 'channel-0',
      '4': [BN],
      '5': [BN],
      __length__: 6,
      from: '0xcBED645B1C1a6254f1149Df51d3591c6B3803007',
      to: '0x00731540cd6060991D6B9C57CE295998d9bC2faB',
      sourcePort: 'transfer',
      sourceChannel: 'channel-0',
      timeoutHeight: [BN],
      amount: [BN]
    }
  }
]
```

## console log
![スクリーンショット 2023-03-22 10 18 28](https://user-images.githubusercontent.com/126585618/226779159-ecd0e88c-0a06-4bad-9446-900631b84514.png)
```
#1: not found
#2: not found
#3: failed to execute a transaction: &{{[] 0 103711 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [] 0xa6f5d48f4bcad58b49876c1295572d7c68e5e3c35b6a081ed3324ab59adf4035 0x0000000000000000000000000000000000000000 103711 0xa03e1466c6087c4a13707288079f2c615506db78cae1ef1d87c18934613338b1 811 0} []}
```

dongrichain commented:
デバッグログ仕込んでログを出してみる
```
2023/03/23 11:29:20 - No acknowledgements to relay between [ibc0]port{transfer} and [ibc1]port{transfer}
----- yui-ibc-solidity.pkg.relay.ethereum.tx.go Send(msgs []sdk.Msg) ------
msgs: [%!v(PANIC=String method: reflect.Value.Interface: cannot return value obtained from unexported field or method) packet:<sequence:1 source_port:"transfer" source_channel:"channel-0" destination_port:"transfer" destination_channel:"channel-0" data:"\0102\022\024\313\355d[\034\032bT\361\024\235\365\0355\221\306\263\2000\007\032\024\000s\025@\315``\231\035k\234W\316)Y\230\331\274/\253" timeout_height:<> > proof_commitment:"Q\350Y\357\352\002\0200\034\254\344=\200\327-\347D\376!&\375\177}!U5\000+1I4\366" proof_height:<revision_height:1 > signer:"cosmos1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqnrql8a" ]
----- SendMsgs ---
opts:  &{0xa89F47C6b463f74d87572b058427dA0A13ec5425 <nil> 0x100e344a0 <nil> <nil> 6382056 <nil>}
msg:  %!v(PANIC=String method: reflect.Value.Interface: cannot return value obtained from unexported field or method)
Error:  <nil>
----- SendMsgs ---
opts:  &{0xa89F47C6b463f74d87572b058427dA0A13ec5425 <nil> 0x100e344a0 <nil> <nil> 6382056 <nil>}
msg:  packet:<sequence:1 source_port:"transfer" source_channel:"channel-0" destination_port:"transfer" destination_channel:"channel-0" data:"\0102\022\024\313\355d[\034\032bT\361\024\235\365\0355\221\306\263\2000\007\032\024\000s\025@\315``\231\035k\234W\316)Y\230\331\274/\253" timeout_height:<> > proof_commitment:"Q\350Y\357\352\002\0200\034\254\344=\200\327-\347D\376!&\375\177}!U5\000+1I4\366" proof_height:<revision_height:1 > signer:"cosmos1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqnrql8a"
Error:  <nil>
2023/03/23 11:29:29 ethereum: failed to send: All attempts fail:
```

siburu commented:
@dongrichain これは `msg.String()` を実行して発生しているエラーですが `recover` されているため処理には影響が無いエラーだと考えてます

dongrichain commented:
```
=== GetTransactionReceipt ===
rs:  <nil>
recoverable:  true
err:  not found
=== GetTransactionReceipt ===
```

dongrichain commented:
txの中身を取得
```
fmt.Println("------------------------ get transaction ------------------------------")
var tx *gethtypes.Transaction
err = cl.rpcClient.Call(&tx, "eth_getTransactionByHash", txHash)
if err != nil {
	fmt.Printf("Error calling eth_getTransactionByHash: %v", err)
}
fmt.Printf("Transaction details:\n")
fmt.Printf("  Hash:       %s\n", tx.Hash().Hex())
fmt.Printf("  Nonce:      %d\n", tx.Nonce())
fmt.Printf("  Gas Price:  %d\n", tx.GasPrice().Uint64())
fmt.Printf("  Gas:        %d\n", tx.Gas())
fmt.Printf("  To:         %s\n", tx.To().Hex())
fmt.Printf("  Value:      %d\n", tx.Value().Uint64())
fmt.Printf("  Input:      %x\n", tx.Data())
fmt.Println("------------------------ get transaction ------------------------------")
```
ログ出力
```
tx hash:  0xcad35271b27b715d291fa78e3e4cf83d0b7876e7989013fa667ad97dce9047bb
=== GetTransactionReceipt ===
rc:  <nil>
recoverable:  true
err:  not found
=== GetTransactionReceipt ===
tx hash:  0xcad35271b27b715d291fa78e3e4cf83d0b7876e7989013fa667ad97dce9047bb
=== GetTransactionReceipt ===
rc:  <nil>
recoverable:  true
err:  not found
=== GetTransactionReceipt ===
tx hash:  0xcad35271b27b715d291fa78e3e4cf83d0b7876e7989013fa667ad97dce9047bb
------------------------ get transaction ------------------------------
Transaction details:
  Hash:       0xcad35271b27b715d291fa78e3e4cf83d0b7876e7989013fa667ad97dce9047bb
  Nonce:      25
  Gas Price:  0
  Gas:        6382056
  To:         0xff77D90D6aA12db33d3Ba50A34fB25401f6e4c4F
  Value:      0
  Input:      236ebd700000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000001a000000000000000000000000000000000000000000000000000000000000001e0000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000087472616e7366657200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096368616e6e656c2d30000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000087472616e7366657200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096368616e6e656c2d300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e08321214cbed645b1c1a6254f1149df51d3591c6b38030071a1400731540cd6060991d6b9c57ce295998d9bc2fab000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002051e859efea0210301cace43d80d72de744fe2126fd7f7d215535002b314934f6
------------------------ get transaction ------------------------------
```
input dataをdecodeする
https://www.npmjs.com/package/ethereum-input-data-decoder
minitoken/solidity/build/IBCHandler.json から ABI 取得
デコードした結果
```
{
  "method": "recvPacket",
  "types": [
    "(tuple,bytes,tuple)"
  ],
  "inputs": [
    [
      [
        {
          "type": "BigNumber",
          "hex": "0x01"
        },
        "transfer",
        "channel-0",
        "transfer",
        "channel-0",
        "0x08321214cbed645b1c1a6254f1149df51d3591c6b38030071a1400731540cd6060991d6b9c57ce295998d9bc2fab",
        [
          {
            "type": "BigNumber",
            "hex": "0x00"
          },
          {
            "type": "BigNumber",
            "hex": "0x00"
          }
        ],
        {
          "type": "BigNumber",
          "hex": "0x00"
        }
      ],
      "0x51e859efea0210301cace43d80d72de744fe2126fd7f7d215535002b314934f6",
      [
        {
          "type": "BigNumber",
          "hex": "0x00"
        },
        {
          "type": "BigNumber",
          "hex": "0x01"
        }
      ]
    ]
  ],
  "names": [
    [
      "msg_",
      [
        "packet",
        "proof",
        "proofHeight"
      ]
    ]
  ]
}
```

dongrichain commented:
https://speee.slack.com/archives/C01TURAASV6/p1679643268714309

<img width="765" alt="スクリーンショット 2023-03-24 19 39 16" src="https://user-images.githubusercontent.com/126585618/227499589-7cb350eb-45e1-4d42-9480-97604cef3b2e.png">

レシート取得できること確認
<img width="783" alt="スクリーンショット 2023-03-24 19 35 08" src="https://user-images.githubusercontent.com/126585618/227499692-5896e955-04c8-4981-a3eb-02480bcb348e.png">

bobにトークン転送されたこと確認
<img width="661" alt="スクリーンショット 2023-03-24 19 35 49" src="https://user-images.githubusercontent.com/126585618/227499765-21c60ffa-c95d-4cdc-97e2-b66f40919421.png">
